### PR TITLE
[onnxifi] Throttle input concurrency with fixed number of threads on Glow side, configured per backend.

### PR DESCRIPTION
**Pull request by rdzhabarov**
*Description*:
This patch should allow handling of concurrent execution of inferences on the caller side.
On the Glow side, we limit concurrency by a number of threads in a thread pool per backend.
This PR should allow running multithreaded C2 execution on a single Glow thread.

*Testing*:
onnxifi integration test

*Documentation*:
N/A

### CODE REVIEW

**Comment by rdzhabarov** [comment]
@yinghai @SplitInfinity This is to fix concurrent execution through onnxifi.
This PR allows Glow backend to run on a single thread no matter of input (C2->onnxifi->glow) concurrency. Once @gcatron makes a progress on the concurrent execution of compiled model on Glow side, we could increase the number of workers to hardware::concurrency for CPU.

**Comment by rdzhabarov** [comment]
@bertmaher mind taking a look?

Mostly it's limiting concurrency and performing setIO/run atomically.

File: lib/Onnxifi/Base.cpp [hunk 1]
```diff
@@ -60,20 +60,46 @@ onnxStatus Graph::initGraph(const void *onnxModel, size_t onnxModelSize,
...
+  llvm::DenseMap<Placeholder *, onnxPointer> outputPlaceholderToBuffer =
+      outputPlaceholderToBuffer_;
+
+  // Once inputs are copied we can allow processing concurrent requests.
+  inputRunMutex_.unlock();
+
+  // Submit graph for asynchronous execution.
+  // Concurrency for executing 'run' method is limited by number of threads in
+  // the backend specific thread pool.
+  backend()->runAsync([inputEvent, outputEvent, inputPlaceholderToBuffer,
```
Comment (yinghai) [inline lib/Onnxifi/Base.cpp]: Is it good to capture these maps by value?

**Comment by rdzhabarov** [comment]
@yinghai that should be tested through the onnxifi driver tests, setIO/run should go through.

**Comment by yinghai** [comment]
Do we have a multithreaded test?

File: lib/Onnxifi/Base.cpp (hunk 1 above)
Comment (rdzhabarov) [reply lib/Onnxifi/Base.cpp]: That's copied intentionally so that those maps internal to graph class are not shared.

**Comment by rdzhabarov** [comment]
Probably not, so some part is not going to be exercised (proper locking specifically).
One way is to run a concurrent test for resnet50 (not sure how hard would be to modify that) or benchmark one (easier to setup).

That's not going to be trivial to add test support right here.

**Comment by yinghai** [comment]
> Sorry for the delayed review -- this looks pretty good so I'll accept to unblock you. I'm a bit nervous about holding a lock from onnxSetGraphIO to onnxRunGraph. But I'm also not sure I understand the onnxifi interface well enough to say whether it's a real problem. Happy to land this now and iterate later, though.

This a bit hacky but I think it gets the job done for now. We actually need `onnxRunGraph(inputDescriptors, outputDescripors)` To solve this problem properly. @Maratyszcza

**Pull request merged**